include/hcl/concurrent/skiplist/skiplist_random/Skiplist.h

```diff
@@ -0,0 +1,732 @@
+        !candidate->markedForRemoval();
+  }
+
+  int findInsertionPointGetMaxLayer(const value_type& data,NodeType* preds[],NodeType* succs[],int* max_layer) const 
+  {
+    *max_layer = maxLayer();
+    return findInsertionPoint(head_.load(std::memory_order_acquire), *max_layer, data, preds, succs);
+  }
+
+  std::pair<NodeType*, int> findNode(const value_type& data) const 
+  {
+    return findNodeDownRight(data);
+  }
+
+  std::pair<NodeType*, int> findNodeDownRight(const value_type& data) const 
+  {
+    NodeType* pred = head_.load(std::memory_order_acquire);
+    int ht = pred->height();
+    NodeType* node = nullptr;
+
+    bool found = false;
+    while (!found) 
+    {
+      for (; ht > 0 && less(data, node = pred->skip(ht - 1)); --ht) 
+      {
+      }
+      if (ht == 0) 
+      {
+        return std::make_pair(node, 0); 
+      }
+      --ht;
+
+      while (greater(data, node)) 
+      {
+        pred = node;
+        node = node->skip(ht);
+      }
+      found = !less(data, node);
+    }
+    return std::make_pair(node, found);
+  }
+
+  std::pair<NodeType*, int> findNodeRightDown(const value_type& data) const 
+  {
+    NodeType* pred = head_.load(std::memory_order_acquire);
+    NodeType* node = nullptr;
+    auto top = maxLayer();
+    int found = 0;
+    for (int layer = top; !found && layer >= 0; --layer) 
+    {
+      node = pred->skip(layer);
+      while (greater(data, node)) 
+      {
+        pred = node;
+        node = node->skip(layer);
+      }
+      found = !less(data, node);
+    }
+    return std::make_pair(node, found);
+  }
+
+  NodeType* lower_bound(const value_type& data) const 
+  {
+    auto node = findNode(data).first;
+    while (node != nullptr && node->markedForRemoval()) 
+    {
+      node = node->skip(0);
+    }
+    return node;
+  }
+
+  void growHeight(int height) 
+  {
+    NodeType* oldHead = head_.load(std::memory_order_acquire);
+    if (oldHead->height() >= height) 
+    {
+      return;
+    }
+
+    NodeType* newHead = recycler_.pop(height,true);
+        //NodeType::create(recycler_.alloc(), height, value_type(), true);
+
+    { 
+      bool g = oldHead->acquireGuard();
+      newHead->copyHead(oldHead);
+      NodeType* expected = oldHead;
+      if (!head_.compare_exchange_strong(
+              expected, newHead, std::memory_order_release)) 
+      {
+        NodeType::destroy(recycler_.alloc(), newHead);
+	oldHead->releaseGuard();
+        return;
+      }
+      oldHead->setMarkedForRemoval();
+      oldHead->releaseGuard();
+    }
+    recycler_.push(oldHead->height(),oldHead);
+    //recycle(oldHead);
+  }
+
+  void recycle(NodeType* node) { recycler_.add(node); }
+
+  NodeRecycler<NodeType, NodeAlloc> recycler_;
+  std::atomic<NodeType*> head_;
+  std::atomic<size_t> size_{0};
+};
+
+template <typename T, typename Comp, typename NodeAlloc, int MAX_HEIGHT>
+class ConcurrentSkipList<T, Comp, NodeAlloc, MAX_HEIGHT>::Accessor 
```

Comment:
So as i understand, you have the main data structures defined here and then built HCL container for these. Do the following with these classes, 
- split the two classes in two files which comments on which class does what in detail.
- put files in include/hcl/base/containers/
- keep naming convention similar to other classes. class names and function names are not consistent.

Reply:
Ok

Reply:
Could you check if the naming conventions are okay now?